demos/demo_stacking_simulation.py

```diff
@@ -0,0 +1,1143 @@
+
+		dir_x = dir_x
+		dir_y = dir_y
+		dir_z = dir_z
+
+		self.rotation_matrix = np.matrix([dir_x, dir_y, dir_z]).T
+		self.quat = pinocchio.Quaternion(self.rotation_matrix).coeffs()
+
+		# Setting the position offset
+		offset =  np.array([0,-0.0054, 0]).reshape(-1,1)
+
+		box_com_position = box_com_position + self.rotation_matrix.dot(offset)
+
+		# if self.t % 100 == 0:
+		# 	print("Block Position:", box_com_position)
+
+		# self.quat[:] = 0
+		# self.quat[-1] = 1
+		# ipdb.set_trace()
+		# print("Box State:", np.concatenate([box_com_position, self.quat]))
+
+		return np.concatenate([box_com_position, self.quat])
+
+
+		ipdb.set_trace()
+
+
+
+	def get_cube_state(self):
+		if self.mode == "simulation":
+			return self.finger.get_block_state()
+
+		if self.mode == "real":
+			quat, trans = self.object_tracker.get_object_pose()
+			quat = np.concatenate([quat[1:],[quat[0]]], axis = 0)
+			quat[:] = 0
+			quat[3] = 1
+			# trans[:2] = 0
+			# trans[2] = 0.05 + 0.2 * (self.t/1000)
+			return self.box_state_from_tip_position()
+			
+			# return np.concatenate([trans, quat], axis=0).reshape(-1,1)
+			# return self.cube.state.copy()[:7]
+
+		if self.mode == "pinocchio":
+			return self.cube.state.copy()[:7]
+
+
+
+	def add_new_object(self, size, position, orientation=[0,0,0,1]):
+		'''
+		Adds a new object to manipulate
+		'''
+		self.object_size = size
+		self.object_position = position
+		self.object_orientation = orientation
+
+		if self.mode == "simulation" or self.mode == "pinocchio": # adding object in pybullet
+			self.finger.import_interaction_objects(size, position, orientation)
+			self.cube = CubeSimulator(dt=1.e-3, int_steps=1)
+			self.dt = 1.e-3
+			# Initial State
+			x_init = np.zeros([self.cube.n,1])
+			x_init[:3] = np.resize(position,(3,1)) 
+			x_init[3:7] = np.resize(orientation,(4,1)) # quaternion
+			self.cube.reset_state(x_init)
+
+			# we can also display it in gepetto viewer 
+			try:
+				self.cube.robot.initViewer(loadModel=True)
+			except:
+				raise BaseException('Gepetto viewer not initialized ! ')
+
+			self.cube.robot.display(x_init[:self.cube.nq])
+
+
+		if self.mode == "real":
+			self.cube = CubeSimulator(dt=1.e-3, int_steps=1)
+			self.dt = 1.e-3
+
+			# Initial State
+			x_init = np.zeros([self.cube.n,1])
+			x_init[:3] = np.resize(position,(3,1)) 
+			x_init[3:7] = np.resize(orientation,(4,1)) # quaternion
+			self.cube.reset_state(x_init)
+
+			initial_object_in_base_link = tf.compose_matrix(angles=(0, 0, 0),
+			                                        translate=(0, 0, 0.05))
+			self.object_tracker = ObjectTracker()
+			self.object_tracker.initialize(initial_object_in_base_link)
+
+		self.cube_q1 = self.get_cube_state()
+		self.cube_q_last = self.get_cube_state()
+
+
+	def get_desired_state(self):
+		# Calculating the desird position to reach
+		self.x_des = [np.matrix([self.object_position[0], self.object_position[1] + self.object_size/2, self.object_position[2] + 0.004]).T,
```

Comment:
These are the desired positions of the three tips while in contact, right?
Is it intentional that the first tip is 4mm higher than the others?  If yes, you need to consider this, when computing the object orientation based on the tip positions.